labs/03_neural_recsys/Explicit_Feedback_Neural_Recommender_System.ipynb

```diff
@@ -288,25 +288,30 @@
    "source": [
     "# For each sample we input the integer identifiers\n",
     "# of a single user and a single item\n",
-    "user_id_input = Input(shape=[1], name='user')\n",
-    "item_id_input = Input(shape=[1], name='item')\n",
-    "\n",
-    "embedding_size = 30\n",
-    "user_embedding = Embedding(output_dim=embedding_size, input_dim=max_user_id + 1,\n",
-    "                           input_length=1, name='user_embedding')(user_id_input)\n",
-    "item_embedding = Embedding(output_dim=embedding_size, input_dim=max_item_id + 1,\n",
-    "                           input_length=1, name='item_embedding')(item_id_input)\n",
-    "\n",
-    "# reshape from shape: (batch_size, input_length, embedding_size)\n",
-    "# to shape: (batch_size, input_length * embedding_size) which is\n",
-    "# equal to shape: (batch_size, embedding_size)\n",
-    "user_vecs = Flatten()(user_embedding)\n",
-    "item_vecs = Flatten()(item_embedding)\n",
-    "\n",
-    "y = Dot(axes=1)([user_vecs, item_vecs])\n",
-    "# y = merge([user_vecs, item_vecs], mode=dot_mode, output_shape=(1,))\n",
-    "\n",
-    "model = Model(inputs=[user_id_input, item_id_input], outputs=y)\n",
+    "class RegressionModel(Model):\n",
+    "    def __init__(self, embedding_size, max_user_id, max_item_id):\n",
+    "        super().__init__()\n",
+    "        \n",
+    "        self.user_embedding = Embedding(output_dim=embedding_size, input_dim=max_user_id + 1,\n",
+    "                                        input_length=1, name='user_embedding')\n",
+    "        self.item_embedding = Embedding(output_dim=embedding_size, input_dim=max_item_id + 1,\n",
+    "                                        input_length=1, name='item_embedding')\n",
+    "        \n",
+    "        # The following two layers don't have parameters.\n",
+    "        self.flatten = Flatten()\n",
+    "        self.dot = Dot(axes=1)\n",
```

Comment:
Is it really useful to store layers without parameters as attribute on the model? Would instantiating those classes in the call function add overhead? If so wouldn't it be possible to use functions instead? I am not familiar with TF2 keras good practices, especially with the Model class API.

Reply:
Assuming we are in eager mode (default), there would be an overhead. And unfortunatly Tensorflow doesn't provide a "functional api" (like PyTorch) of this layer.

It seems that the Google-recommended way is instantiating in the __ini__ (or build) any kind of layers, even those without parameters.